File: lib/iam-construct.ts
import { Construct } from 'constructs';
import * as aws from '@cdktf/provider-aws';

export interface IamConstructProps {
  environment: string;
  commonTags: { [key: string]: string };
}

export class IamConstruct extends Construct {
  public readonly ec2RoleArn: string;
  public readonly ec2ProfileName: string;
  public readonly s3ServiceRoleArn: string;
  public readonly cloudwatchRoleArn: string;

  constructor(scope: Construct, id: string, config: IamConstructProps) {
    super(scope, id);

    new aws.provider.AwsProvider(this, 'aws', {
      region: process.env.AWS_REGION || 'us-west-2',
    });

    const ec2Role = new aws.iamRole.IamRole(this, 'Ec2Role', {
      name: `${config.environment}-ec2-role`,
      assumeRolePolicy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Principal: { Service: 'ec2.amazonaws.com' },
            Action: 'sts:AssumeRole',
          },
        ],
      }),
      tags: config.commonTags,
    });

    const ec2Profile = new aws.iamInstanceProfile.IamInstanceProfile(
      this,
      'Ec2Profile',
      {
        name: `${config.environment}-ec2-profile`,
        role: ec2Role.name,
        tags: config.commonTags,
      }
    );

    new aws.iamRolePolicy.IamRolePolicy(this, 'Ec2Policy', {
      name: `${config.environment}-ec2-policy`,
      role: ec2Role.id,
      policy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Action: [
              'cloudwatch:PutMetricData',
              'ec2:DescribeVolumes',
              'ec2:DescribeTags',
              'logs:PutLogEvents',
              'logs:CreateLogGroup',
              'logs:CreateLogStream',
              'logs:DescribeLogStreams',
              'logs:DescribeLogGroups',
            ],
            Resource: '*',
          },
          {
            Effect: 'Allow',
            Action: ['s3:GetObject', 's3:PutObject'],
            Resource: `arn:aws:s3:::${config.environment}-*/*`,
          },
        ],
      }),
    });

    const s3ServiceRole = new aws.iamRole.IamRole(this, 'S3ServiceRole', {
      name: `${config.environment}-s3-service-role`,
      assumeRolePolicy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Principal: { Service: 's3.amazonaws.com' },
            Action: 'sts:AssumeRole',
          },
        ],
      }),
      tags: config.commonTags,
    });

    const cloudwatchRole = new aws.iamRole.IamRole(this, 'CloudWatchRole', {
      name: `${config.environment}-cloudwatch-role`,
      assumeRolePolicy: JSON.stringify({
        Version: '2012-10-17',
        Statement: [
          {
            Effect: 'Allow',
            Principal: { Service: 'logs.amazonaws.com' },
            Action: 'sts:AssumeRole',
          },
        ],
      }),
      tags: config.commonTags,
    });

    this.ec2RoleArn = ec2Role.arn;
    this.ec2ProfileName = ec2Profile.name;
    this.s3ServiceRoleArn = s3ServiceRole.arn;
    this.cloudwatchRoleArn = cloudwatchRole.arn;
  }
}

File: archive/cdktf-java/Pr3276/lib/src/main/java/app/constructs/IAMConstruct.java
package app.constructs;

import app.config.Config;
import com.hashicorp.cdktf.providers.aws.iam_role.IamRole;
import com.hashicorp.cdktf.providers.aws.iam_role.IamRoleConfig;
import com.hashicorp.cdktf.providers.aws.iam_role_policy.IamRolePolicy;
import com.hashicorp.cdktf.providers.aws.iam_role_policy.IamRolePolicyConfig;
import com.hashicorp.cdktf.providers.aws.iam_role_policy_attachment.IamRolePolicyAttachment;
import com.hashicorp.cdktf.providers.aws.iam_role_policy_attachment.IamRolePolicyAttachmentConfig;
import software.constructs.Construct;

public class IAMConstruct extends Construct {
    private final IamRole codePipelineRole;
    private final IamRole codeBuildRole;
    private final IamRole codeDeployRole;

    public IAMConstruct(final Construct scope, final String id, final Config config,
                        final ArtifactStorageConstruct artifactStorage) {
        super(scope, id);

        // CodePipeline Service Role
        this.codePipelineRole = new IamRole(this, "codepipeline-role",
                IamRoleConfig.builder()
                        .name(config.resourceName(config.projectName() + "-codepipeline-role"))
                        .assumeRolePolicy("""
                                {
                                    "Version": "2012-10-17",
                                    "Statement": [{
                                        "Effect": "Allow",
                                        "Principal": {
                                            "Service": "codepipeline.amazonaws.com"
                                        },
                                        "Action": "sts:AssumeRole"
                                    }]
                                }
                                """)
                        .build());

        new IamRolePolicy(this, "codepipeline-policy",
                IamRolePolicyConfig.builder()
                        .role(codePipelineRole.getName())
                        .name(config.projectName() + "-codepipeline-policy")
                        .policy(String.format("""
                                        {
                                            "Version": "2012-10-17",
                                            "Statement": [
                                                {
                                                    "Effect": "Allow",
                                                    "Action": [
                                                        "s3:GetObject",
                                                        "s3:GetObjectVersion",
                                                        "s3:PutObject",
                                                        "s3:GetBucketVersioning"
                                                    ],
                                                    "Resource": [
                                                        "%s",
                                                        "%s/*"
                                                    ]
                                                },
                                                {
                                                    "Effect": "Allow",
                                                    "Action": [
                                                        "s3:GetObject",
                                                        "s3:GetObjectVersion",
                                                        "s3:GetBucketVersioning",
                                                        "s3:GetBucketAcl",
                                                        "s3:GetBucketLocation",
                                                        "s3:GetObjectTagging",
                                                        "s3:GetObjectVersionTagging",
                                                        "s3:ListBucket"
                                                    ],
                                                    "Resource": [
                                                        "%s",
                                                        "%s/*"
                                                    ]
                                                },
                                                {
                                                    "Effect": "Allow",
                                                    "Action": [
                                                        "codebuild:BatchGetBuilds",
                                                        "codebuild:StartBuild"
                                                    ],
                                                    "Resource": "*"
                                                },
                                                {
                                                    "Effect": "Allow",
                                                    "Action": [
                                                        "codedeploy:CreateDeployment",
                                                        "codedeploy:GetApplication",
                                                        "codedeploy:GetApplicationRevision",
                                                        "codedeploy:GetDeployment",
                                                        "codedeploy:GetDeploymentConfig",
                                                        "codedeploy:RegisterApplicationRevision"
                                                    ],
                                                    "Resource": "*"
                                                },
                                                {
                                                    "Effect": "Allow",
                                                    "Action": [
                                                        "kms:Decrypt",
                                                        "kms:GenerateDataKey"
                                                    ],
                                                    "Resource": "%s"
                                                },
                                                {
                                                    "Effect": "Allow",
                                                    "Action": [
                                                        "sns:Publish"
                                                    ],
                                                    "Resource": "*"
                                                }
                                            ]
                                        }
                                        """,
                                artifactStorage.getArtifactBucket().getArn(),
                                artifactStorage.getArtifactBucket().getArn(),
                                artifactStorage.getSourceBucket().getArn(),
                                artifactStorage.getSourceBucket().getArn(),
                                artifactStorage.getKmsKey().getArn()))
                        .build());

        // CodeBuild Service Role
        this.codeBuildRole = new IamRole(this, "codebuild-role",
                IamRoleConfig.builder()
                        .name(config.resourceName(config.projectName() + "-codebuild-role"))
                        .assumeRolePolicy("""
                                {
                                    "Version": "2012-10-17",
                                    "Statement": [{
                                        "Effect": "Allow",
                                        "Principal": {
                                            "Service": "codebuild.amazonaws.com"
                                        },
                                        "Action": "sts:AssumeRole"
                                    }]
                                }
                                """)
                        .build());

        new IamRolePolicy(this, "codebuild-policy",
                IamRolePolicyConfig.builder()
                        .role(codeBuildRole.getName())
                        .name(config.projectName() + "-codebuild-policy")
                        .policy(String.format("""
                                        {
                                            "Version": "2012-10-17",
                                            "Statement": [
                                                {
                                                    "Effect": "Allow",
                                                    "Action": [
                                                        "logs:CreateLogGroup",
                                                        "logs:CreateLogStream",
                                                        "logs:PutLogEvents"
                                                    ],
                                                    "Resource": "arn:aws:logs:*:*:*"
                                                },
                                                {
                                                    "Effect": "Allow",
                                                    "Action": [
                                                        "s3:GetObject",
                                                        "s3:GetObjectVersion",
                                                        "s3:PutObject"
                                                    ],
                                                    "Resource": [
                                                        "%s/*"
                                                    ]
                                                },
                                                {
                                                    "Effect": "Allow",
                                                    "Action": [
                                                        "kms:Decrypt",
                                                        "kms:GenerateDataKey"
                                                    ],
                                                    "Resource": "%s"
                                                }
                                            ]
                                        }
                                        """,
                                artifactStorage.getArtifactBucket().getArn(),
                                artifactStorage.getKmsKey().getArn()))
                        .build());

        // CodeDeploy Service Role
        this.codeDeployRole = new IamRole(this, "codedeploy-role",
                IamRoleConfig.builder()
                        .name(config.resourceName(config.projectName() + "-codedeploy-role"))
                        .assumeRolePolicy("""
                                {
                                    "Version": "2012-10-17",
                                    "Statement": [{
                                        "Effect": "Allow",
                                        "Principal": {
                                            "Service": "codedeploy.amazonaws.com"
                                        },
                                        "Action": "sts:AssumeRole"
                                    }]
                                }
                                """)
                        .build());

        new IamRolePolicyAttachment(this, "codedeploy-policy-attachment",
                IamRolePolicyAttachmentConfig.builder()
                        .role(codeDeployRole.getName())
                        .policyArn("arn:aws:iam::aws:policy/service-role/AWSCodeDeployRole")
                        .build());
    }

    public IamRole getCodePipelineRole() {
        return codePipelineRole;
    }

    public IamRole getCodeBuildRole() {
        return codeBuildRole;
    }

    public IamRole getCodeDeployRole() {
        return codeDeployRole;
    }
}
